relayer/chain/substrate/outgoing_extrinsics.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020 Snowfork
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package substrate
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/snowfork/go-substrate-rpc-client/v2/types"
+	"golang.org/x/sync/errgroup"
+)
+
+const MaxWatchedExtrinsics = 500
+
+type extrinsicPool struct {
+	sync.Mutex
+	conn     *Connection
+	eg       *errgroup.Group
+	log      *logrus.Entry
+	maxNonce uint32
+	watched  uint
+}
+
+func newExtrinsicPool(eg *errgroup.Group, conn *Connection, log *logrus.Entry) *extrinsicPool {
+	ep := extrinsicPool{
+		conn:    conn,
+		eg:      eg,
+		log:     log,
+		watched: 0,
+	}
+	return &ep
+}
+
+func (ep *extrinsicPool) WaitForSubmitAndWatch(ctx context.Context, nonce uint32, ext *types.Extrinsic) {
+	defer ep.Unlock()
+
+	for {
+		ep.Lock()
+		if ep.hasCapacity() {
+			ep.eg.Go(func() error {
+				return ep.submitAndWatchLoop(ctx, nonce, ext)
+			})
+
+			ep.watched++
+			return
+		}
+		ep.Unlock()
+
+		time.Sleep(100 * time.Millisecond)
+	}
```

Comment:
Instead of a loop with a sleep, the more idiomatic way would be to use a buffered channel of size `MaxWatchedExtrinsics`. No need to change that, just noticed it ;)